src/libsyntax_pos/hygiene.rs

```diff
@@ -129,13 +138,98 @@ impl HygieneData {
         }
         HYGIENE_DATA.with(|data| f(&mut *data.borrow_mut()))
     }
+
+    pub fn safe_with<T, F: FnOnce(&HygieneData) -> T>(f: F) -> T {
+        // FIXME(twk): not sure how this would behave...
```

Comment:
Okay, a lot of symptoms point to having two thread locals being the issue. However, replacing them with one leads to absolutely uncontrolled memory consumption.